tensorflow/compiler/xla/service/cpu/cpu_runtime_avx.h

```diff
@@ -38,6 +38,7 @@ typedef float V8F32AVX __attribute__((__vector_size__(32)));
 
 extern "C" {
 
+#ifdef __AVX__
```

Comment:
I thought the existing code would work as:

 - The definition would not be compiled in since it is guarded by `__AVX__`
 - This symbol would resolve to null (since we're weak linking)
 - We'd do the right thing based on what `GetAvailableIntrinsics` returned

Is any of the steps above not working as intended?

Reply:
Yeah weak links are weird with clang on OSX. 

Here is a reduced test case:
```
/* weaklink.c */
#include <stdio.h>
int weakfn() __attribute__((weak));

// Uncomment next line to test a strong link
//int weakfn() { return 0xCAFE; }

int main()
{
    if (weakfn) {
        printf ("Weak link found.. 0x%x\n", weakfn());
        return weakfn();
    } else {
        printf ("No Weak link found..");
        return 0;
    }
}
```
The above will work well with gcc or clang on linux but will fail with clang on OSX (both Apple's and clang upstream) with the following error:
```
Undefined symbols for architecture x86_64:
  "_weakfn", referenced from:
      _main in weaklink-4baf7d.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation).
```
Our compile currently also fails with the same kind of message:
```
Undefined symbols for architecture x86_64:
  "__xla_cpu_runtime_ExpV8F32AVX(float vector[8])", referenced from:
      xla::cpu::SimpleOrcJIT::SimpleOrcJIT(llvm::TargetOptions const&, llvm::CodeGenOpt::Level, bool, bool, bool, std::__1::function<tensorflow::Status (llvm::Module const&)>, std::__1::function<tensorflow::Status (llvm::Module const&)>) in libsimple_orc_jit.a(simple_orc_jit.o)
      xla::cpu::(anonymous namespace)::RegisterKnownJITSymbols() in libsimple_orc_jit.a(simple_orc_jit.o)
  "__xla_cpu_runtime_LogV8F32AVX(float vector[8])", referenced from:
      xla::cpu::(anonymous namespace)::RegisterKnownJITSymbols() in libsimple_orc_jit.a(simple_orc_jit.o)
  "___xla_cpu_runtime_ExpV4F32NEON", referenced from:
      xla::cpu::SimpleOrcJIT::SimpleOrcJIT(llvm::TargetOptions const&, llvm::CodeGenOpt::Level, bool, bool, bool, std::__1::function<tensorflow::Status (llvm::Module const&)>, std::__1::function<tensorflow::Status (llvm::Module const&)>) in libsimple_orc_jit.a(simple_orc_jit.o)
      xla::cpu::(anonymous namespace)::RegisterKnownJITSymbols() in libsimple_orc_jit.a(simple_orc_jit.o)
  "___xla_cpu_runtime_LogV4F32NEON", referenced from:
      xla::cpu::(anonymous namespace)::RegisterKnownJITSymbols() in libsimple_orc_jit.a(simple_orc_jit.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Target //tensorflow/compiler/aot:tfcompile failed to build
```

So I wrapped the registrations in the header file with \_\_AVX\_\_ and friends. Open to any other way to solve it.

Reply:
On a related note the extern C also causes heartburn 
```
./tensorflow/compiler/xla/service/cpu/cpu_runtime_neon.h:55:30: warning: '__xla_cpu_runtime_ExpV4F32NEON' has C-linkage specified, but returns incomplete type 'xla::cpu::runtime::V4F32NEON' which could be incompatible with C [-Wreturn-type-c-linkage]
xla::cpu::runtime::V4F32NEON __xla_cpu_runtime_ExpV4F32NEON(
```

Reply:
Ok, please mention this discrepancy in weak linkage in the commit message.